signer/signer.go

```diff
@@ -0,0 +1,262 @@
+// containing the information about the signature.
+func (s *SignatureEnvelope) Verify(trustedCerts []*x509.Certificate) (*x509.Certificate, *SignerInfo, error) {
+	if len(s.rawSignatureEnvelope) == 0 {
+		return nil, nil, SignatureNotFoundError{}
+	}
+
+	if len(trustedCerts) == 0 {
+		return nil, nil, MalformedArgumentError{param: "trustedCerts"}
+	}
+
+	integrityError := s.internalEnvelope.validateIntegrity()
+	if integrityError != nil {
+		return nil, nil, integrityError
+	}
+
+	singerInfo, singerInfoErr := s.GetSignerInfo()
+	if singerInfoErr != nil {
+		return nil, nil, singerInfoErr
+	}
+
+	certChain := singerInfo.CertificateChain
+	err := nx509.ValidateCertChain(certChain)
+	if err != nil {
+		return nil, nil, singerInfoErr
+	}
+
+	resolvedTrust, err := validateAuthenticity(certChain, trustedCerts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return resolvedTrust, singerInfo, nil
+}
+
+// Sign generates Signature using given SignRequest.
+func (s *SignatureEnvelope) Sign(req SignRequest) ([]byte, error) {
+	if err := validateSignRequest(req); err != nil {
+		return nil, err
+	}
+
+	sig, err := s.internalEnvelope.signPayload(req)
+	if err != nil {
+		return nil, err
+	}
+
+	s.rawSignatureEnvelope = sig
+	return sig, err
+}
+
+// GetSignerInfo returns information about the Signature envelope
+func (s SignatureEnvelope) GetSignerInfo() (*SignerInfo, error) {
+	if len(s.rawSignatureEnvelope) == 0 {
+		return nil, SignatureNotFoundError{}
+	}
+
+	signInfo, err := s.internalEnvelope.getSignerInfo()
+	if err != nil {
+		return nil, MalformedSignatureError{msg: fmt.Sprintf("signature envelope format is malformed. error: %s", err.Error())}
+	}
+
+	if err := validateSignerInfo(signInfo); err != nil {
+		return nil, err
+	}
+	return signInfo, nil
+}
+
+// validateSignerInfo performs basic set of validations on SignerInfo struct.
+func validateSignerInfo(info *SignerInfo) error {
+	if len(info.Payload) == 0 {
+		return MalformedSignatureError{msg: "payload not present"}
+	}
+
+	// TODO: perform PayloadContentType value validations
+	if len(info.PayloadContentType) == 0 {
+		return MalformedSignatureError{msg: "signature content type not present or is empty"}
+	}
+
+	signTime := info.SignedAttributes.SigningTime
+	if signTime.IsZero() {
+		return MalformedSignatureError{msg: "singing-time not present"}
+	}
+
+	expTime := info.SignedAttributes.Expiry
+	if !expTime.IsZero() {
+		if expTime.Before(signTime) || expTime.Equal(signTime) {
+			return MalformedSignatureError{msg: "expiry cannot be equal or before the signing time"}
+		}
+	}
+
+	if len(info.CertificateChain) == 0 {
+		return MalformedSignatureError{msg: "certificate-chain not present or is empty"}
+	}
+
+	if len(info.Signature) == 0 {
+		return MalformedSignatureError{msg: "signature not present or is empty"}
+	}
+
+	return nil
+}
+
+// validateSignRequest performs basic set of validations on SignRequest struct.
+func validateSignRequest(req SignRequest) error {
```

Comment:
It seems that there are duplicated codes in `validateSignerInfo` and `validateSignRequest`. Can we refactor them?

Reply:
Done